OptimizelySDK/Bucketing/DecisionService.cs

```diff
@@ -253,5 +246,156 @@ public void SaveVariation(Experiment experiment, Variation variation, UserProfil
+
+            // For all rules before the everyone else rule
+            for (int i=0; i < rolloutRulesLength - 1; i++)
+            {
+                var rolloutRule = rollout.Experiments[i];
+
+                if (ExperimentUtils.IsUserInExperiment(ProjectConfig, rolloutRule, filteredAttributes))
+                {
+                    Logger.Log(LogLevel.DEBUG, $"Attempting to bucket user \"{userId}\" into rollout rule \"{rolloutRule.Key}\".");
+                    variation = Bucketer.Bucket(ProjectConfig, rolloutRule, bucketingId, userId);
+
+                    if (variation == null)
+                    {
+                        Logger.Log(LogLevel.DEBUG, $"User \"{userId}\" is excluded due to traffic allocation. Checking \"Eveyrone Else\" rule now.");
+                        break;
+                    }
+
+                    return variation;
+                }
+                else
+                {
+                    Logger.Log(LogLevel.DEBUG, $"User \"{userId}\" does not meet the audience conditions to be in rollout rule \"{rolloutRule.Key}\".");
+                }
+            }
+
+            // Bucket the user into the last rule which is everyone else rule
+            var everyoneElseRolloutRule = rollout.Experiments[rolloutRulesLength - 1];
+            variation = Bucketer.Bucket(ProjectConfig, everyoneElseRolloutRule, bucketingId, userId);
+
+            if (variation == null)
+            {
+                Logger.Log(LogLevel.DEBUG, $"User \"{userId}\" is excluded from \"Everyone Else\" rule for feature flag \"{featureFlag.Key}\".");
+            }
+
+            return variation;
+        }
+
+        /// <summary>
+        /// Get the variation if the user is bucketed for one of the experiments on this feature flag.
+        /// </summary>
+        /// <param name = "featureFlag" >The feature flag the user wants to access.</param>
+        /// <param name = "userId" >User Identifier</param>
+        /// <param name = "filteredAttributes" >The user's attributes. This should be filtered to just attributes in the Datafile.</param>
+        /// <returns>null if the user is not bucketed into the rollout or if the feature flag was not attached to a rollout.
+        /// otherwise the Variation the user is bucketed into</returns>
+        public virtual Variation GetVariationForFeatureExperiment(FeatureFlag featureFlag, string userId, UserAttributes filteredAttributes)
+        {
+            if (featureFlag.ExperimentIds == null || featureFlag.ExperimentIds.Count == 0)
+            {
+                Logger.Log(LogLevel.INFO, $"The feature flag \"{featureFlag.Key}\" is not used in any experiments.");
+                return null;
+            }
+
+            foreach (var experimentId in featureFlag.ExperimentIds)
+            {
+                var experiment = ProjectConfig.GetExperimentFromId(experimentId);
+
+                if (string.IsNullOrEmpty(experiment.Key))
```

Comment:
if `experiment` is null, will calling `experiment.Key` throw a NPE. Why not check if `experiment` is not null?

Reply:
In case when experiment Id doesn't get found, we do not return null from ProjectConfig.GetExperimentFromId method but a new Experiment entity containing empty key. So the NPE never gets thrown in this case.